README.md
xface-extension-calendar
==============


RELEASENOTES.md
<!--
#
# Copyright 2012-2013, Polyvi Inc. (http://polyvi.github.io/openxface)
# This program is distributed under the terms of the GNU General Public License.
# 
# This file is part of xFace.
# 
# xFace is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# xFace is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with xFace.  If not, see <http://www.gnu.org/licenses/>.
#
-->

# Release Notes
### 1.0.0 Tue Jan 07 2014 16:08:06 GMT+0800 (CST)
 *  [android]Modify source codes package path
 *  added jasmine tests unique id
### 1.0.1 Mon Jan 27 2014 15:56:03 GMT+0800 (CST)
 *  batch modify .reviewboard
 *  [android]Sync xFace3.1 calendar plugin
 *  Incremented plugin version on dev branch to 1.0.1-dev

test/autotest/tests/calendar.tests.js
describe('Calendar (xFace.ui.Calendar)', function () {
	/**define some constants*/
	var MAX_YEARS    = 2100;
	var MAX_MONTHS   = 12;
	var MAX_DAYS     = 31;
	var MAX_HOURS    = 23;
	var MAX_MINUTES  = 59;
	var MIN_YEARS    = 1900;
	var MIN_MONTHS   = 1;
	var MIN_DAYS     = 1;
	var MIN_HOURS    = 0;
	var MIN_MINUTES  = 0;

    it("calendar.spec.1 should exist", function() {
        expect(xFace.ui.Calendar).toBeDefined();
    });

    it("calendar.spec.2 should contain a getTime function", function() {
        expect(xFace.ui.Calendar.getTime).toBeDefined();
        expect(typeof xFace.ui.Calendar.getTime == 'function').toBe(true);
    });

    it("calendar.spec.3 should contain a getDate function", function() {
        expect(xFace.ui.Calendar.getDate).toBeDefined();
        expect(typeof xFace.ui.Calendar.getDate == 'function').toBe(true);
    });

    describe("getTime method", function() {
        it("calendar.spec.4 it should be  error  with hour over 23", function() {
            var win = jasmine.createSpy(),
                fail = jasmine.createSpy().andCallFake(function() {
                });
			var hour = 24;
			var minute = 14;
            runs(function () {
                xFace.ui.Calendar.getTime(win, fail, hour, minute);
            });
            waitsFor(function () { return fail.wasCalled; }, "fail never called", Tests.TEST_TIMEOUT);
            runs(function () {
                expect(win).not.toHaveBeenCalled();
            });
        });

        it("calendar.spec.5 it should be  error  with hour less 0", function() {
            var win = jasmine.createSpy(),
                fail = jasmine.createSpy().andCallFake(function() {
                });
			var hour = -1;
			var minute = 14;
            runs(function () {
                xFace.ui.Calendar.getTime(win, fail, hour, minute);
            });
            waitsFor(function () { return fail.wasCalled; }, "fail never called", Tests.TEST_TIMEOUT);
            runs(function () {
                expect(win).not.toHaveBeenCalled();
            });
        });

        it("calendar.spec.6 it should be  error  with minutes over 59", function() {
            var win = jasmine.createSpy(),
                fail = jasmine.createSpy().andCallFake(function() {
                });
			var hour = 22;
			var minute = 60;
            runs(function () {
                xFace.ui.Calendar.getTime(win, fail, hour, minute);
            });
            waitsFor(function () { return fail.wasCalled; }, "fail never called", Tests.TEST_TIMEOUT);
            runs(function () {
                expect(win).not.toHaveBeenCalled();
            });
        });

        it("calendar.spec.7 it should be  error  with minutes less 0", function() {
            var win = jasmine.createSpy(),
                fail = jasmine.createSpy().andCallFake(function() {
                });
			var hour = 22;
			var minute = -1;
            runs(function () {
                xFace.ui.Calendar.getTime(win, fail, hour, minute);
            });
            waitsFor(function () { return fail.wasCalled; }, "fail never called", Tests.TEST_TIMEOUT);
            runs(function () {
                expect(win).not.toHaveBeenCalled();
            });
        });
    });

    describe("getDate method", function() {
        it("calendar.spec.8 it should be  error  with year over 2100", function() {
            var win = jasmine.createSpy(),
                fail = jasmine.createSpy().andCallFake(function() {
                });
			var year = 2101;
			var month = 11;
			var day = 14;
            runs(function () {
                xFace.ui.Calendar.getDate(win, fail, year, month , day);
            });
            waitsFor(function () { return fail.wasCalled; }, "fail never called", Tests.TEST_TIMEOUT);
            runs(function () {
                expect(win).not.toHaveBeenCalled();
            });
        });

        it("calendar.spec.9 it should be  error  with year less 1900", function() {
            var win = jasmine.createSpy(),
                fail = jasmine.createSpy().andCallFake(function() {
                });
			var year = 1899;
			var month = 11;
			var day = 14;
            runs(function () {
                xFace.ui.Calendar.getDate(win, fail, year, month , day);
            });
            waitsFor(function () { return fail.wasCalled; }, "fail never called", Tests.TEST_TIMEOUT);
            runs(function () {
                expect(win).not.toHaveBeenCalled();
            });
        });

        it("calendar.spec.10 it should be  error  with month over 12", function() {
            var win = jasmine.createSpy(),
                fail = jasmine.createSpy().andCallFake(function() {
                });
			var year = 2012;
			var month = 13;
			var day = 14;
            runs(function () {
                xFace.ui.Calendar.getDate(win, fail, year, month , day);
            });
            waitsFor(function () { return fail.wasCalled; }, "fail never called", Tests.TEST_TIMEOUT);
            runs(function () {
                expect(win).not.toHaveBeenCalled();
            });
        });

        it("calendar.spec.11 it should be  error  with month less 1", function() {
            var win = jasmine.createSpy(),
                fail = jasmine.createSpy().andCallFake(function() {
                });
			var year = 2012;
			var month = 0;
			var day = 14;
            runs(function () {
                xFace.ui.Calendar.getDate(win, fail, year, month , day);
            });
            waitsFor(function () { return fail.wasCalled; }, "fail never called", Tests.TEST_TIMEOUT);
            runs(function () {
                expect(win).not.toHaveBeenCalled();
            });
        });

        it("calendar.spec.12 it should be  error  with day over 31", function() {
            var win = jasmine.createSpy(),
                fail = jasmine.createSpy().andCallFake(function() {
                });
			var year = 2012;
			var month = 11;
			var day = 32;
            runs(function () {
                xFace.ui.Calendar.getDate(win, fail, year, month , day);
            });
            waitsFor(function () { return fail.wasCalled; }, "fail never called", Tests.TEST_TIMEOUT);
            runs(function () {
                expect(win).not.toHaveBeenCalled();
            });
        });

        it("calendar.spec.13 it should be  error  with day less 1", function() {
            var win = jasmine.createSpy(),
                fail = jasmine.createSpy().andCallFake(function() {
                });
			var year = 2012;
			var month = 11;
			var day = 0;
            runs(function () {
                xFace.ui.Calendar.getDate(win, fail, year, month , day);
            });
            waitsFor(function () { return fail.wasCalled; }, "fail never called", Tests.TEST_TIMEOUT);
            runs(function () {
                expect(win).not.toHaveBeenCalled();
            });
        });
    });
});


src/ios/XCalendarExt.h

/*
 Copyright 2012-2013, Polyvi Inc. (http://polyvi.github.io/openxface)
 This program is distributed under the terms of the GNU General Public License.

 This file is part of xFace.

 xFace is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 xFace is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with xFace.  If not, see <http://www.gnu.org/licenses/>.
 */

//
//  XCalendarExt.h
//  xFaceLib
//
//

#import <Cordova/CDVPlugin.h>

@interface XCalendarExt : CDVPlugin
{
}

/**
    初始化时间控件显示的时间并通过时间控件获取系统的时间.
    @param command.arguments
    - 0 hour 初始设置时间控件显示的小时 (iOS 不支持设置初始时间，默认为系统当前时间)
    - 1 minute 初始设置时间控件显示的分钟
 */
- (void) getTime:(CDVInvokedUrlCommand*)command;

/**
    初始化日期控件显示的日期通过日期控件获取系统的日期.
    @param command.arguments
    - 0 year 初始设置日期控件显示的年份 (iOS 不支持设置初始时间，默认为系统当前时间)
    - 1 month 初始设置日期控件显示的月份
    - 2 day 初始设置日期控件显示的天
 */
- (void) getDate:(CDVInvokedUrlCommand*)command;

@end


src/ios/XCalendarExt.m

/*
 Copyright 2012-2013, Polyvi Inc. (http://polyvi.github.io/openxface)
 This program is distributed under the terms of the GNU General Public License.

 This file is part of xFace.

 xFace is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 xFace is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with xFace.  If not, see <http://www.gnu.org/licenses/>.
 */

//
//  XCalendarExt.m
//  xFaceLib
//
//
#import "XCalendarExt.h"
#import "XCalendarExt_Privates.h"

#import <Cordova/CDVAvailability.h>
#import <Cordova/CDVInvokedUrlCommand.h>
#import <Cordova/CDVPluginResult.h>
#import <Cordova/NSArray+Comparisons.h>
#import <Cordova/CDVDebug.h>

#define VIEW_DEFAULT_WIDTH          320
#define VIEW_DEFAULT_HEIGHT         480
#define TOOLBAR_HEIGHT              44
#define POPOVER_CONTENT_WIDTH       320
#define POPOVER_CONTENT_HEIGHT      244


@implementation XCalendarExt

@synthesize popoverController;

- (void) getTime:(CDVInvokedUrlCommand*)command
{
    callbackId = command.callbackId;
    [self showDatePikcer];
    [datePicker setDatePickerMode:UIDatePickerModeTime];
    [datePicker setDate:[NSDate date]];
}

- (void) getDate:(CDVInvokedUrlCommand*)command
{
    callbackId = command.callbackId;
    [self showDatePikcer];
    [datePicker setDatePickerMode:UIDatePickerModeDate];
    [datePicker setDate:[NSDate date]];
}

- (void)dateChanged:(id)sender
{
    NSDate *date = ((UIDatePicker*)sender).date;
    DLog(@"dateChanged:%@", date);
}

- (void)doCancelClick:(id)sender
{
    //取消 datePickerView
    if (popoverController)
    {
        [popoverController dismissPopoverAnimated:YES];
        popoverController.delegate = nil;
    }

    if (action)
    {
        [action dismissWithClickedButtonIndex:0 animated:YES];
    }

    DLog(@"datepicker canceled !!");
    CDVPluginResult *result = [CDVPluginResult resultWithStatus: CDVCommandStatus_NO_RESULT];
    [self.commandDelegate sendPluginResult:result callbackId:callbackId];
}

- (void)doDoneClick:(id)sender
{
    //取消datePickerView 把设置好的时间返回JS端
    if (popoverController)
    {
        [popoverController dismissPopoverAnimated:YES];
        popoverController.delegate = nil;
    }

    if (action)
    {
        [action dismissWithClickedButtonIndex:1 animated:YES];
    }

    NSDate *selected = [datePicker date];
    DLog(@"you selected date:%@", selected);

    CDVPluginResult *result = [CDVPluginResult resultWithStatus: CDVCommandStatus_OK messageAsDictionary:
                                [self getDateComponentsFrom:selected]];
    [self.commandDelegate sendPluginResult:result callbackId:callbackId];
}

-(NSDictionary*) getDateComponentsFrom:(NSDate*)date
{
    NSCalendar *cal = [NSCalendar currentCalendar];
    unsigned int unitFlags = NSYearCalendarUnit | NSMonthCalendarUnit | NSDayCalendarUnit |NSHourCalendarUnit | NSMinuteCalendarUnit;
    NSDateComponents *dd = [cal components:unitFlags fromDate:date];

    NSDictionary *dateComponents = [NSDictionary dictionaryWithObjectsAndKeys:
                                   [NSString stringWithFormat:@"%d", [dd year]], @"year",
                                   [NSString stringWithFormat:@"%d", [dd month]], @"month",
                                   [NSString stringWithFormat:@"%d", [dd day]], @"day",
                                   [NSString stringWithFormat:@"%d", [dd hour]], @"hour",
                                   [NSString stringWithFormat:@"%d", [dd minute]], @"minute",
                                   nil];
    return dateComponents;
}

- (void)createDatePickerView
{
    datePickerView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, VIEW_DEFAULT_WIDTH, VIEW_DEFAULT_HEIGHT)];
    datePickerView.backgroundColor = [UIColor whiteColor];

    datePicker = [[UIDatePicker alloc] initWithFrame:CGRectMake(0.0, TOOLBAR_HEIGHT, 0.0, 0.0)];
    [datePicker addTarget:self action:@selector(dateChanged:) forControlEvents:UIControlEventValueChanged];

    UIToolbar *datePickerToolbar = [[UIToolbar alloc] initWithFrame:CGRectMake(0, 0, VIEW_DEFAULT_WIDTH, TOOLBAR_HEIGHT)];
    datePickerToolbar.barStyle = UIBarStyleBlackOpaque;

    NSMutableArray *barItems = [[NSMutableArray alloc] init];
    UIBarButtonItem *btnFlexibleSpace = [[UIBarButtonItem alloc]
                                         initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace
                                         target:self action:nil];
    [barItems addObject:btnFlexibleSpace];

    UIBarButtonItem *btnCancel = [[UIBarButtonItem alloc]
                                  initWithBarButtonSystemItem:UIBarButtonSystemItemCancel
                                  target:self
                                  action:@selector(doCancelClick:)];
    [barItems addObject:btnCancel];

    UIBarButtonItem *btnDone = [[UIBarButtonItem alloc]
                                initWithBarButtonSystemItem:UIBarButtonSystemItemDone
                                target:self
                                action:@selector(doDoneClick:)];

    [barItems addObject:btnDone];
    [datePickerToolbar setItems:barItems animated:YES];

    [datePickerView addSubview: datePickerToolbar];
    [datePickerView addSubview: datePicker];
}

- (void)showDatePikcer
{
    if(CDV_IsIPad())
    {
        //iPad
        if (nil == popoverController)
        {
            [self createDatePickerView];
            UIViewController *popoverContent = [[UIViewController alloc] init];
            popoverContent.view = datePickerView;
            popoverContent.contentSizeForViewInPopover = CGSizeMake(POPOVER_CONTENT_WIDTH, POPOVER_CONTENT_HEIGHT);
            popoverController = [[UIPopoverController alloc] initWithContentViewController:popoverContent];
        }
        [self.popoverController presentPopoverFromRect:CGRectMake(0,0,VIEW_DEFAULT_WIDTH,VIEW_DEFAULT_HEIGHT)
                                                inView:[self.viewController.view superview]
                              permittedArrowDirections:UIPopoverArrowDirectionAny
                                              animated:YES];
    }
    else
    {
        //iphone
        if (nil == action)
        {
            [self createDatePickerView];
            action = [[UIActionSheet alloc] initWithTitle:@""
                                                 delegate:nil
                                        cancelButtonTitle:nil
                                   destructiveButtonTitle:nil
                                        otherButtonTitles:nil];
            [action addSubview:datePickerView];
        }

        UIWindow* appWindow = [self.viewController.view window];
        [action showInView: appWindow];
        [action setBounds:CGRectMake(0, 0, VIEW_DEFAULT_WIDTH, VIEW_DEFAULT_HEIGHT)];
    }
}

@end


src/android/XCalendarExt.java

/*
 Copyright 2012-2013, Polyvi Inc. (http://polyvi.github.io/openxface)
 This program is distributed under the terms of the GNU General Public License.

 This file is part of xFace.

 xFace is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 xFace is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with xFace.  If not, see <http://www.gnu.org/licenses/>.
*/

package com.polyvi.xface.extension.calendar;

import java.util.Calendar;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.app.Activity;
import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.widget.DatePicker;
import android.widget.TimePicker;

public class XCalendarExt extends CordovaPlugin {

    private static final String COMMAND_GET_TIME = "getTime";
    private static final String COMMAND_GET_DATE = "getDate";
    private CallbackContext mCallbackContext;

    @Override
    public boolean execute(String action, JSONArray args,
            CallbackContext callbackContext) throws JSONException {
        Calendar calendar = Calendar.getInstance();
        mCallbackContext = callbackContext;
        if (COMMAND_GET_TIME.equals(action)) {
            // 若没有输入，则使用当前时间，反之亦然
            int hours = calendar.get(Calendar.HOUR_OF_DAY);
            int minutes = calendar.get(Calendar.MINUTE);
            if (2 == args.length()) {
                hours = args.getInt(0);
                minutes = args.getInt(1);
            }
            getTime(hours, minutes);
        } else if (COMMAND_GET_DATE.equals(action)) {
            // 若没有输入，则使用当前日期，反之亦然
            int year = calendar.get(Calendar.YEAR);
            int month = calendar.get(Calendar.MONTH) + 1;
            int day = calendar.get(Calendar.DAY_OF_MONTH);
            if (3 == args.length()) {
                year = args.getInt(0);
                month = args.getInt(1);
                day = args.getInt(2);
            }
            getDate(year, month, day);
        }
        PluginResult result = new PluginResult(PluginResult.Status.NO_RESULT);
        result.setKeepCallback(true);
        callbackContext.sendPluginResult(result);
        return true;
    }

    /**
     * 初始化时间控件显示的时间并通过时间控件获取系统的时间
     *
     * @param hour
     *            初始设置时间控件显示的小时
     * @param minute
     *            初始设置时间控件显示的分钟
     */
    private void getTime(final int hour, final int minute) {
        ((Activity) cordova.getActivity()).runOnUiThread(new Runnable() {
            @Override
            public void run() {
                new TimePickerDialog(cordova.getActivity(),
                        new TimePickerDialog.OnTimeSetListener() {
                            @Override
                            public void onTimeSet(TimePicker view,
                                    int hourOfDay, int minute) {
                                setTime(hourOfDay, minute);
                            }
                        }, hour, minute, true).show();
            }
        });
    }

    /**
     * 初始化日期控件显示的日期通过日期控件获取系统的日期
     *
     * @param year
     *            初始设置日期控件显示的年份
     * @param month
     *            初始设置日期控件显示的月份
     * @param day
     *            初始设置日期控件显示的天
     */
    private void getDate(final int year, final int month, final int day) {
        ((Activity) cordova.getActivity()).runOnUiThread(new Runnable() {
            @Override
            public void run() {
                new DatePickerDialog(cordova.getActivity(),
                        new DatePickerDialog.OnDateSetListener() {
                            @Override
                            public void onDateSet(DatePicker view, int year,
                                    int monthOfYear, int dayOfMonth) {
                                setDate(year, monthOfYear, dayOfMonth);
                            }
                        }, year, (month - 1), day).show();// 由于系统计算月份是以0开始，所以系统在日期选择器
                                                          // 控件中会自动把设置的初始月份加1，为了保证月份准确
                                                          // 显示，所以这里将js传来的月份数减1
            }
        });
    }

    /**
     * 根据时间控件选择的时间结果，设置需要返回给js的时间
     *
     * @param hour
     *            通过时间控件设置的小时
     * @param minute
     *            通过时间控件设置的分钟
     */
    private void setTime(int hour, int minute) {
        JSONObject time = new JSONObject();
        try {
            time.put("hour", hour);
            time.put("minute", minute);
            mCallbackContext.success(time);
        } catch (JSONException e) {
            PluginResult result = new PluginResult(PluginResult.Status.ERROR);
            mCallbackContext.sendPluginResult(result);
        }
    }

    /**
     * 根据日期控件选择的日期结果，设置需要返回给js的日期
     *
     * @param year
     *            通过日期控件设置的年份
     * @param month
     *            通过日期控件设置的月份
     * @param day
     *            通过日期控件设置的天
     */
    private void setDate(int year, int month, int day) {
        JSONObject date = new JSONObject();
        try {
            date.put("year", year);
            date.put("month", (month + 1));// 由于系统返回的月份是以0开始，所以在返回给js月份的时候加1
            date.put("day", day);
            mCallbackContext.success(date);
        } catch (JSONException e) {
            PluginResult result = new PluginResult(PluginResult.Status.ERROR);
            mCallbackContext.sendPluginResult(result);
        }
    }
}


src/ios/private/XCalendarExt_Privates.h

/*
 Copyright 2012-2013, Polyvi Inc. (http://polyvi.github.io/openxface)
 This program is distributed under the terms of the GNU General Public License.

 This file is part of xFace.

 xFace is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 xFace is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with xFace.  If not, see <http://www.gnu.org/licenses/>.
 */

//
//  XCalendarExt_Privates.h
//  xFaceLib
//
//

#import <Foundation/Foundation.h>
#import "XCalendarExt.h"

@interface XCalendarExt ()
{
    UIView *datePickerView;
    UIActionSheet *action;
    UIDatePicker *datePicker;
    NSString      *callbackId;
}

@property (strong) UIPopoverController* popoverController;

/**
    创建DatePickerView.
 */
- (void) createDatePickerView;

/**
    显示DatePickerView.
 */
- (void) showDatePikcer;

/**
    获取Date的各组成部分,year,month,day等.
    @param date 被分解的date对象
    @return 返回一个year,month,day等组成的NSDictionary对象
 */
-(NSDictionary*) getDateComponentsFrom:(NSDate*)date;

@end


www/Calendar.js
/*
 Copyright 2012-2013, Polyvi Inc. (http://www.xface3.com)
 This program is distributed under the terms of the GNU General Public License.

 This file is part of xFace.

 xFace is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 xFace is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with xFace.  If not, see <http://www.gnu.org/licenses/>.
 */


/**
 * ui模块提供系统原生控件，方便应用直接使用
 * @module ui
 */
var argscheck = require('cordova/argscheck'),
    exec = require('cordova/exec');

/**
 * 此类提供系统原生calendar控件支持.此类不能通过new来创建相应的对象,只能通过xFace.ui.Calendar
 * 对象来直接使用该类中定义的方法(Android,iOS,WP8)
 * @class Calendar
 * @static
 * @platform Android, iOS, WP8
 * @since 3.0.0
 */
function Calendar() {}

/**定义时间的一些常量*/
var MAX_YEARS    = 2100;
var MAX_MONTHS   = 12;
var MAX_DAYS     = 31;
var MAX_HOURS    = 23;
var MAX_MINUTES  = 59;
var MIN_YEARS    = 1900;
var MIN_MONTHS   = 1;
var MIN_DAYS     = 1;
var MIN_HOURS    = 0;
var MIN_MINUTES  = 0;

/**
 * 打开原生时间控件.可以指定控件显示的初始时间,如果用户不传入初始时间，则默认为当前系统时间.(Android,iOS,WP8)
 * 注意：初始时间要么不传，要么全传，否则会报错。
 * @example
        //通过Calendar控件获取用户选取的时间
        function getTime(){
            xFace.ui.Calendar.getTime(
                function(res){
                    alert(res.hour);
                    alert(res.minute);
                },
                function(){alert(" Calendar fail!");}
                );
        }
 * @method getTime
 * @param {Function} successCallback   成功的回调函数，返回用户设置的时间.
 * @param {Object}  successCallback.obj  回调函数的参数为一个带有hour,minute属性的Object对象
 * @param {Function} [errorCallback]     失败的回调函数
 * @param {Number} [hours]   初始小时值(iOS,WP8上不支持传参初始化Calendar控件,默认显示系统当前的时间)
 * @param {Number} [minutes] 初始分钟值(iOS,WP8上不支持传参初始化,不需要该参数）
 * @platform Android,iOS,WP8
 * @since 3.0.0
 */
Calendar.prototype.getTime = function(successCallback, errorCallback, hours, minutes) {
    argscheck.checkArgs('fFNN', 'Calendar.getTime', arguments);
    if(arguments.length == 3){
        if(errorCallback && typeof errorCallback == "function") {
            errorCallback("The parameter length is invalid! ");
            return;
        }
    }
    var checkTime = function(hour, minute) {
        if((hour < MIN_HOURS) || (minute < MIN_MINUTES) ||
            (hour > MAX_HOURS) || (minute > MAX_MINUTES)){
            return false;
        }
        //实例一个Date对象
        var d = new Date();
        d.setHours(hour);
        d.setMinutes(minute);
        return (d.getHours() == hour &&
        d.getMinutes() == minute);
    };
    var newArguments = [];
    if(arguments.length == 4) {
        if(!checkTime(hours, minutes)){
            if(errorCallback && typeof errorCallback == "function") {
                errorCallback("The parameter value is invalid! ");
                return;
            }
        } else {
            newArguments = [hours, minutes];
        }
    }
    exec(successCallback, errorCallback,"Calendar", "getTime", newArguments);
};

/**
 * 打开原生日期控件。可以指定控件显示的初始日期,如果用户不传入初始日期，则默认为当前系统日期.(Android,iOS,WP8)
 * 注意：初始日期要么不传，要么全传，否则会报错。
 * @example
        //通过Calendar控件获取用户选取的日期
        function getDate(){
            xFace.ui.Calendar.getDate(
                function(res){
                    alert(res.year);
                    alert(res.month);
                    alert(res.day);
                },
                function(){alert(" Calendar fail!");}
                    2012,09,10 );
        }

 * @method getDate
 * @param {Function} successCallback   成功回调函数，返回用户设置的日期.
 * @param {Object}  successCallback.obj  回调函数的参数为一个带有year,month,day属性的Object对象
 * @param {Function} [errorCallback]      失败回调函数
 * @param {Number} [year]    初始年值(iOS,WP8上不支持传参初始化Calendar控件,默认显示系统当前的日期)
 * @param {Number} [month]   初始月份值(iOS,WP8上不支持传参初始化,不需要该参数)
 * @param {Number} [day]     初始日值(iOS,WP8上不支持传参初始化,不需要该参数)
 * @platform Android,iOS,WP8
 * @since 3.0.0
 */
Calendar.prototype.getDate = function(successCallback, errorCallback, year, month, day) {
    argscheck.checkArgs('fFNNN', 'Calendar.getDate', arguments);
    if(arguments.length != 5 && arguments.length != 2 && arguments.length != 1){
        if(errorCallback && typeof errorCallback == "function") {
            errorCallback("The parameter length is invalid! ");
            return;
        }
    }
    var checkDate = function(years, months, days) {
        if((years < MIN_YEARS) || (months < MIN_MONTHS) || (days < MIN_DAYS) ||
            (years > MAX_YEARS) || (months > MAX_MONTHS) || (days > MAX_DAYS) ){
            return false;
        }
        //实例一个Date对象并初始化各个属性值，注意月份是从0开始，因此减1
        var d = new Date(years, months-1, days);
        //判断输入时期是否合法 ，同样月份需要加1
        return (d.getFullYear() == years && d.getMonth()+1 == months && d.getDate()== days);
    };
    var newArguments = [];
    if(arguments.length == 5) {
        if(!checkDate(year,month,day)){
            if(errorCallback && typeof errorCallback == "function") {
                errorCallback("The parameter value is invalid! ");
                return;
            }
        } else {
            newArguments = [year, month, day];
        }
    }

    exec(successCallback, errorCallback,"Calendar", "getDate",newArguments);
};

module.exports = new Calendar();